### PR TITLE
hswidget: refactoring state encoding system

### LINKED ISSUE #331: hswidget: widgetState encoder needs refactor
a current implementation of state encoder (#230) is really scruffy. If we want to add a new field to the `widgetState` struct, we also need to add it to Encode/Decode methods.

### PR DESCRIPTION
**Pull request by gucio321**
Hi there,
#230 works in a bit strange way, so I decided to refactor it here

### CODE REVIEW

**Comment by gucio321** [comment]
Sadly, I need to export some `state` fields, because json.Marshal only considers exported fields.
https://golang.org/pkg/encoding/json/#Marshal
> Struct values encode as JSON objects. Each **exported** struct field becomes a member of the object, using the field name as the object key, unless the field is omitted for one of the reasons given below.

**Comment by gravestench** [comment]
please resolve conflicts.

**Review by gravestench — CHANGES_REQUESTED** [review]
This is almost 2K lines of code being edited in this PR.

I need you to describe, in a lot more detail than you did in your first comment, what you are doing in this PR, why you are doing it, and why you think we should merge this into `master`.

**Comment by gucio321** [comment]
ok, so here is some changelog:
- change the way, editors state is encoded.
in #230 I was encoding/decoding state using OpenDiablo2 d2datautils package, but it doesn't look very good. In this pr I rewritten the stuff to use `encoding/json`. The code is much easier to understand now.
the `2K lines of code being edited` are because `encoding/json` skips unexported struct fields so I had to export them

- fix minor capitalization in comments
- remove two unnecessary fields from palettegridwidget.widgetState struct (they were `nolint:unused`ed)

**Comment by gravestench** [comment]
@gucio321 Your previous comment would be great as a commit message.

Going forward, we need to file issues before we start work like this. The previous implementation _was not broken_, so why do we need these changes? 

Please file an issue describing your intent with this PR. Also, there are some merge conflicts.

**Comment by gucio321** [comment]
okey @gravestench I opened an issue. This PR should fix #331

File: hswidget/palettegrideditorwidget/state.go [hunk 1]
```diff
@@ -18,105 +17,32 @@ const (
...
-	sw.PushBytes(ws.g)
-	sw.PushBytes(ws.b)
-	sw.PushBytes(byte(len(ws.hex)))
-	sw.PushBytes([]byte(ws.hex)...)
-
-	return sw.GetBytes()
+	ws.Mode = widgetModeGrid
 }
 
 func (ws *widgetState) Decode(data []byte) {
```
Comment (gravestench) [inline hswidget/palettegrideditorwidget/state.go]: You have removed other such `Decode` methods, why are we keeping this one for the palette grid editor state?

Comment (gucio321) [reply hswidget/palettegrideditorwidget/state.go]: thanks for pointing this, I think I missed it
see c85c329

**Review by gravestench — CHANGES_REQUESTED** [review]
I'm still seeing merge conflicts with upstream for this PR, can you merge/rebase with upstream?

**Comment by gucio321** [comment]
@gravestench done

**Review by gravestench — APPROVED** [review]

**Comment by gravestench** [comment]
Awesome!

So, in short, the reasons for change were:
* state transcoding logic needed to be defined for each widget (doesn't scale!)
* changing a widget's state required changes to the transcoder functions (doesn't scale!)

Your proposed solution is to switch to **json marshaling.** These changes lower the complexity of state transcoding, which makes the code much more legible, scalable, and easy to maintain. This has to do with the fact that **transcode functions need not be defined**; structs only need exported fields with json directives in order to be marshaled.

**Comment by gucio321** [comment]
> Awesome!
> 
> So, in short, the reasons for change were:
> 
>     * state transcoding logic needed to be defined for each widget (doesn't scale!)
> 
>     * changing a widget's state required changes to the transcoder functions (doesn't scale!)
> 
> 
> Your proposed solution is to switch to **json marshaling.** These changes lower the complexity of state transcoding, which makes the code much more legible, scalable, and easy to maintain. This has to do with the fact that **transcode functions need not be defined**; structs only need exported fields with json directives in order to be marshaled.

exactly, I'm sorry if my messages wasn't clear enough

**Pull request merged**